product_contract/models/sale_order_line.py

```diff
@@ -0,0 +1,274 @@
+            "analytic_account_id": self.order_id.analytic_account_id.id,
+        }
+
+    def create_contract_line(self, contract):
+        contract_line_model = self.env["contract.line"]
+        contract_line = self.env["contract.line"]
+        predecessor_contract_line = False
+        for rec in self:
+            if rec.contract_line_id:
+                # If the upsell/downsell line start at the same date or before
+                # the contract line to replace supposed to start, we cancel
+                # the one to be replaced. Otherwise we stop it.
+                if rec.date_start <= rec.contract_line_id.date_start:
+                    # The contract will handel the contract line integrity
+                    # An exception will be raised if we try to cancel an
+                    # invoiced contract line
+                    rec.contract_line_id.cancel()
+                elif (
+                    not rec.contract_line_id.date_end
+                    or rec.date_start <= rec.contract_line_id.date_end
+                ):
+                    rec.contract_line_id.stop(rec.date_start - relativedelta(days=1))
+                    predecessor_contract_line = rec.contract_line_id
+            if predecessor_contract_line:
+                new_contract_line = contract_line_model.create(
+                    rec._prepare_contract_line_values(
+                        contract, predecessor_contract_line.id
+                    )
+                )
+                predecessor_contract_line.successor_contract_line_id = new_contract_line
+            else:
+                new_contract_line = contract_line_model.create(
+                    rec._prepare_contract_line_values(contract)
+                )
+            contract_line |= new_contract_line
+        return contract_line
+
+    @api.constrains("contract_id")
+    def _check_contract_sale_partner(self):
+        for rec in self:
+            if rec.contract_id:
+                if rec.order_id.partner_id != rec.contract_id.partner_id:
+                    raise ValidationError(
+                        _(
+                            "Sale Order and contract should be "
+                            "linked to the same partner"
+                        )
+                    )
+
+    @api.constrains("product_id", "contract_id")
+    def _check_contract_sale_contract_template(self):
+        for rec in self:
+            if rec.contract_id:
+                if (
+                    rec.contract_id.contract_template_id
+                    and rec.contract_template_id != rec.contract_id.contract_template_id
+                ):
+                    raise ValidationError(
+                        _("Contract product has different contract template")
+                    )
+
+    def _compute_invoice_status(self):
+        res = super(SaleOrderLine, self)._compute_invoice_status()
+        for line in self.filtered("contract_id"):
```

Comment:
```suggestion
        self.filtered("contract_id").update({"invoice_status": "no"})
```

Better for call stack